packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,562 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+
+import 'dart:ui' show lerpDouble;
+
+/// A class that abstracts over a Material slice and a Material gap.
+///
+/// All [MergeableMaterialItem] objects need a [LocalKey].
+abstract class MergeableMaterialItem {
+  MergeableMaterialItem(this.key) {
+    assert(key != null);
+  }
+
+  final LocalKey key;
+}
+
+/// A class that can be used as a child to [MergeableMaterial]. It is a slice
+/// of [Material] that animates merging with other slices.
+///
+/// All [MaterialSlice] objects need a [LocalKey].
+class MaterialSlice extends MergeableMaterialItem {
+  /// Creates a slice of [Material] that's mergeable within a
+  /// [MergeableMaterial].
+  MaterialSlice({
+    LocalKey key,
+    this.child
+  }) : super(key);
+
+  /// The widget below this widget in the tree.
+  final Widget child;
+
+  @override
+  String toString() {
+    return 'MergeableSlice(key: $key, child: $child)';
+  }
+}
+
+/// A class that represents a gap within [MergeableMaterial].
+///
+/// All [MaterialGap] objects need a [LocalKey].
+class MaterialGap extends MergeableMaterialItem {
+  /// Creates a Material gap with a given size.
+  MaterialGap({
+    LocalKey key,
+    this.size: 16.0
+  }) : super(key);
+
+  /// The size of the gap, irrespective of the axis used.
+  final double size;
+
+  @override
+  String toString() {
+    return 'MaterialGap(key: $key, child: $size)';
+  }
+}
+
+/// A widget useful for animating the merging and separating of [Material]
+/// slices. The widget takes [MergeableMaterialItem] children, uses a block
+/// layout algorithm, and separates [MaterialSlice] wherever it sees a
+/// [MaterialGap]. Animations work by figuring out where new gaps were
+/// added/removed when the widget gets rebuilt by checking the children's keys,
+/// which is why keys on children are mandatory.
+///
+/// The addition of extra slices is not animated and neither is the adding of
+/// any extra items at the end of the children list.
```

Comment:
maybe add something like "When adding new slices, animate the contents of the slice so that they grow from zero to their natural dimensions, and vice versa before removing a slice."